src/main/java/bio/terra/service/filedata/azure/tables/TableFileDao.java

```diff
@@ -0,0 +1,141 @@
+
+import bio.terra.service.filedata.exception.FileSystemCorruptException;
+import bio.terra.service.filedata.google.firestore.*;
+import com.azure.core.http.rest.PagedIterable;
+import com.azure.data.tables.TableClient;
+import com.azure.data.tables.TableServiceClient;
+import com.azure.data.tables.models.ListEntitiesOptions;
+import com.azure.data.tables.models.TableEntity;
+import com.google.api.core.SettableApiFuture;
+import com.google.common.collect.ImmutableList;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.stereotype.Component;
+
+/**
+ * TableFileDao provides CRUD operations on the "files" storage table in Azure. Objects in the file
+ * table are referred to by the dataset (owner of the files) and by any snapshots that reference the
+ * files. File naming is handled by the directory DAO. This DAO just handles the basic operations
+ * for managing the collection. It does not have logic for protecting against deleting files that
+ * have dependencies or creating files with duplicate paths.
+ */
+@Component
+public class TableFileDao {
+  private final Logger logger = LoggerFactory.getLogger(TableFileDao.class);
+  private final FireStoreUtils fireStoreUtils;
+  private final ExecutorService executor;
+  private final String TABLE_NAME = "files";
+  private final String PARTITION_KEY = "partitionKey";
+  private static final int DELETE_BATCH_SIZE = 500;
+
+  @Autowired
+  TableFileDao(
+      FireStoreUtils fireStoreUtils, @Qualifier("performanceThreadpool") ExecutorService executor) {
+    this.fireStoreUtils = fireStoreUtils;
+    this.executor = executor;
+  }
+
+  public void createFileMetadata(TableServiceClient tableServiceClient, FireStoreFile newFile) {
+    tableServiceClient.createTableIfNotExists(TABLE_NAME);
+    TableClient tableClient = tableServiceClient.getTableClient(TABLE_NAME);
+    TableEntity entity =
+        new TableEntity(PARTITION_KEY, newFile.getFileId())
+            .addProperty("fileId", newFile.getFileId())
+            .addProperty("mimeType", newFile.getMimeType())
+            .addProperty("description", newFile.getDescription())
+            .addProperty("bucketResourceId", newFile.getBucketResourceId())
+            .addProperty("loadTag", newFile.getLoadTag())
+            .addProperty("fileCreatedDate", newFile.getFileCreatedDate())
+            .addProperty("gspath", newFile.getGspath())
+            .addProperty("checksumCrc32c", newFile.getChecksumCrc32c())
+            .addProperty("checksumMd5", newFile.getChecksumMd5())
+            .addProperty("size", newFile.getSize());
+    logger.info("creating file metadata for fileId: " + newFile.getFileId());
+    tableClient.createEntity(entity);
```

Comment:
I _think_ we may want to have the createEntity be transactional.  Can you remind me what we do for Firestore here?  Does it created these entries transactionally?  I suspect not, I just want to double check (I suspect it's more when we do the path stuff)

Reply:
With firestore, we do create the file metadata in a transaction: https://github.com/DataBiosphere/jade-data-repo/blob/develop/src/main/java/bio/terra/service/filedata/google/firestore/FireStoreFileDao.java#L58

Just for my own understanding, what's the value of using a transaction to add one record to an Azure table? The examples from their documentation use transactions to perform multiple operations, so I wasn't sure!

Reply:
If it's for individual writes, it's not a big deal. It's if we're doing a bulk operation that it's useful so we don't end up in a semi-committed state